rosetta/transactor/transactor.go

```diff
@@ -0,0 +1,592 @@
+	}
+	_, err = t.validate.Account(receiver)
+	if err != nil {
+		return identifier.Block{}, 0, nil, nil, fmt.Errorf("invalid receiver account: %w", err)
+	}
+
+	// Validate the reference block identifier.
+	refBlockID := identifier.Block{
+		Hash: tx.ReferenceBlockID.String(),
+	}
+	height, blockID, err := t.validate.Block(refBlockID)
+	if err != nil {
+		return identifier.Block{}, 0, nil, nil, fmt.Errorf("invalid reference block: %w", err)
+	}
+
+	// Create the send operation.
+	sendOp := object.Operation{
+		ID: identifier.Operation{
+			Index: 0,
+		},
+		AccountID: sender,
+		Type:      dps.OperationTransfer,
+		Amount: object.Amount{
+			Value: "-" + amount,
+			Currency: identifier.Currency{
+				Symbol:   dps.FlowSymbol,
+				Decimals: dps.FlowDecimals,
+			},
+		},
+	}
+
+	// Create the receive operation.
+	receiveOp := object.Operation{
+		ID: identifier.Operation{
+			Index: 1,
+		},
+		AccountID: receiver,
+		Type:      dps.OperationTransfer,
+		Amount: object.Amount{
+			Value: amount,
+			Currency: identifier.Currency{
+				Symbol:   dps.FlowSymbol,
+				Decimals: dps.FlowDecimals,
+			},
+		},
+	}
+
+	// Create the operations list.
+	ops := []object.Operation{
+		sendOp,
+		receiveOp,
+	}
+
+	// Since we only support sender as the payer/proposer, we never expect any payload signatures.
+	if len(tx.PayloadSignatures) > 0 {
+		return identifier.Block{}, 0, nil, nil, failure.InvalidSignature{
+			Description: failure.NewDescription("unexpected payload signature found",
+				failure.WithInt("signatures", len(tx.PayloadSignatures))),
+		}
+	}
+
+	// We may be parsing an unsigned transaction - if that's the case, we're done.
+	if len(tx.EnvelopeSignatures) == 0 {
+		return rosettaBlockID(height, blockID), 0, ops, nil, nil
+	}
+
+	// We don't support multiple signatures.
+	if len(tx.EnvelopeSignatures) > 1 {
+		return identifier.Block{}, 0, nil, nil, failure.InvalidSignature{
+			Description: failure.NewDescription("unexpected envelope signatures found",
+				failure.WithInt("signatures", len(tx.EnvelopeSignatures))),
+		}
+	}
+
+	// Validate that it is the sender who signed the transaction.
+	signer := tx.EnvelopeSignatures[0].Address
+	if signer != authorizer {
+		return identifier.Block{}, 0, nil, nil, failure.InvalidSignature{
+			Description: failure.NewDescription("invalid signer account",
+				failure.WithString("have_signer", signer.String()),
+				failure.WithString("want_signer", authorizer.String()),
+				failure.WithString("signature", hex.EncodeToString(tx.EnvelopeSignatures[0].Signature))),
+		}
+	}
+
+	// Check that the signature is valid.
+	address := flow.BytesToAddress(signer[:])
+	err = t.verifySignature(payload, address)
+	if err != nil {
+		return identifier.Block{}, 0, nil, nil, fmt.Errorf("could not verify signature: %w", err)
+	}
+
+	// Create the signers list.
+	signers := []identifier.Account{
+		sender,
+	}
+
+	return rosettaBlockID(height, blockID), tx.ProposalKey.SequenceNumber, ops, signers, nil
```

Comment:
Once again, I really think that returning a struct here would be much much better. Mixing up a multi-return function call with (MANY!) other return parameters here is a code smell IMO.

Reply:
I don't think returning a struct here would be better, even if I agree that this is a code smell. By returning a struct, you break the boundaries between the components, which is imo a bigger mistake than having these ugly many returns. So the solution is neither of them, imo.